packages/catalog-process/src/services/readModelService.ts

```diff
@@ -0,0 +1,289 @@
+const catalog = db.collection("eservices");
+const agreements = db.collection("agreements");
+
+function arrayToFilter<T, F extends object>(
+  array: T[],
+  f: (array: T[]) => F
+): F | undefined {
+  return array.length > 0 ? f(array) : undefined;
+}
+
+export const readModelGateway = {
+  async getCatalogItems(
+    authData: AuthData,
+    eservicesIds: string[],
+    producersIds: string[],
+    states: DescriptorState[],
+    agreementStates: PersistentAgreementState[],
+    offset: number,
+    limit: number,
+    name?: { value: string; exactMatch: boolean }
+  ): Promise<ListResult<CatalogItem>> {
+    const ids = await match(agreementStates.length)
+      .with(0, () => eservicesIds)
+      .otherwise(async () =>
+        (
+          await this.listAgreements(
+            eservicesIds,
+            [authData.organizationId],
+            [],
+            agreementStates
+          )
+        ).map((a) => a.eserviceId)
+      );
+
+    if (agreementStates.length > 0 && ids.length === 0) {
+      return emptyListResult;
+    }
+
+    const nameFilter = name
+      ? {
+          "data.name": {
+            $regex: name.exactMatch ? `^${name.value}$$` : name.value,
+            $options: "i",
+          },
+        }
+      : {};
+
+    const aggregationPipeline = [
+      {
+        $match: {
+          ...nameFilter,
+          ...arrayToFilter(states, (states) => ({
+            "data.descriptors": { $elemMatch: { state: { $in: states } } },
+          })),
+          ...arrayToFilter(ids, (ids) => ({ "data.id": { $in: ids } })),
+          ...arrayToFilter(producersIds, (producersIds) => ({
+            "data.producerId": { $in: producersIds },
+          })),
+        },
+      },
+      {
+        $project: {
+          data: 1,
+          computedColumn: { $toLower: ["$data.name"] },
+        },
+      },
+      {
+        $sort: { computedColumn: 1 },
+      },
+    ];
+
+    const data = await catalog
+      .aggregate([...aggregationPipeline, { $skip: offset }, { $limit: limit }])
+      .toArray();
+
+    const result = z.array(catalogItem).safeParse(data.map((d) => d.data));
+    return result.success
+      ? {
+          results: result.data,
+          totalCount: (
+            await catalog
+              .aggregate([...aggregationPipeline, { $count: "count" }])
+              .toArray()
+          )[0].count as number,
+        }
+      : emptyListResult;
+  },
+  async getCatalogItemById(
+    id: string
+  ): Promise<(CatalogItem & { version: number }) | undefined> {
+    const data = await catalog.findOne(
+      { "data.id": id },
+      { projection: { data: true, metadata: true } }
+    );
+
+    if (data) {
+      const result = z
+        .object({
+          metadata: z.object({ version: z.number() }),
+          data: catalogItem,
+        })
+        .safeParse(data);
+      return result.success
+        ? { ...result.data.data, version: result.data.metadata.version }
+        : undefined;
```

Comment:
as above, I think for all queries we should distinguish the "no result" case (expected) with the "parse error" case (should log the error: I think we're likely going to run into those errors if we evolve the read model e.g. because it doesn't match the existing one now, so a log will help)